web/pandas/pdeps/0013-logical-type-system.md

```diff
@@ -0,0 +1,161 @@
+
+If you wanted to try and be more explicit about using pyarrow, you could use the ``pd.ArrowDtype`` wrapper. But this unfortunately exposes gaps when trying to use that pattern across all backends:
+
+| logical type | NumPy    | pandas extension | pyarrow                           |
+| int          | np.int64 | pd.Int64Dtype()  | pd.ArrowDtype(pa.int64())         |
+| string       | N/A      | pd.StringDtype() | pd.ArrowDtype(pa.string())        |
+| datetime     | N/A      | ???              | pd.ArrowDtype(pa.timestamp("us")) |
+
+It would stand to reason in this approach that you could use a ``pd.DatetimeDtype()`` but no such type exists (there is a ``pd.DatetimeTZDtype`` which requires a timezone).
+
+The third issue is that the extent to which pandas may support any given type is unclear. Issue [#58307](https://github.com/pandas-dev/pandas/issues/58307) highlights one example. It would stand to reason that you could interchangeably use a pandas datetime64 and a pyarrow timestamp, but that is not always true. Another common example is the use of NumPy fixed length strings, which users commonly try to use even though we claim no real support for them (see [#5764](https://github.com/pandas-dev/pandas/issues/57645)).
+
+## Assessing the Current Type System(s)
+
+A best effort at visualizing the current type system(s) with types that we currently "support" or reasonably may want to is shown [in this comment](https://github.com/pandas-dev/pandas/issues/58141#issuecomment-2047763186). Note that this does not include the ["pyarrow_numpy"](https://github.com/pandas-dev/pandas/pull/58451) string data type or the string data type that uses the NumPy 2.0 variable length string data type (see [comment](https://github.com/pandas-dev/pandas/issues/57073#issuecomment-2080798081)) as they are under active discussion.
+
+## Proposal
+
+Derived from the hierarchical visual in the previous section, this PDEP proposes that pandas supports at least all of the following _logical_ types, excluding any type widths for brevity:
+
+  - Signed Integer
+  - Unsigned Integer
+  - Floating Point
+  - Fixed Point
+  - Boolean
+  - Date
+  - Datetime
+  - Duration
+  - Interval
+  - Period
+  - Binary
+  - String
+  - Dictionary
+  - List
+  - Struct
+
+To ensure we maintain all of the current functionality of our existing type system(s), a base type structure would need to look something like:
+
+```python
+class BaseType:
+
+    @property
+    def dtype_backend -> Literal["pandas", "numpy", "pyarrow"]:
+        """
+        Library is responsible for the array implementation
+        """
+        ...
+
+    @property
+    def physical_type:
```

Comment:
are there use cases for this other than the pyarrow strings?  i.e. is it really needed for the general case?

Reply:
makes sense, thanks.  is there a reason to include the decimalN case but not (int|float|uint)N?  is that distinct from itemsize?

its still unclear to me why this is needed in the general case.  the way it is presented here suggests this is something i access on my dtype object, so in the DatetimeTZDtype case i would check dtype.physical_type.tz instead of dtype.tz?  or i could imagine standardizing the dtype constructors so you would write `dtype = pd.datetime(physical_type=something_that_specifices_tz)` instead of `dtype=pd.datetime(tz=tz)`.  Neither of these seems great, so i feel like I'm missing something.

Reply:
That's a good question about the decimal case. If coming from a database world its not uncommon to have to pick between different integer sizes, but I think rare to pick between different decimal sizes. With decimals you would usually just select your precision, from which we _could_ dispatch to a 128 or 256 bit implementation...but yea I'm not sure the best way to approach that

Doing something like `pd.datetime(physical_type=something_that_specifies_tz)` is definitely not what this PDEP suggests users should do. This PDEP suggests the 99% use case would be `pd.datetime(tz=tz)` and a user should not care if we use pandas or pyarrow underneath that.

Using https://github.com/pandas-dev/pandas/issues/58315 as an example again, if we used the `ser.dt.date` accessor on a series with a logical type of `pd.datetime()`, then the user should expect back a `pd.date()` back. That gives us a lot of flexibility as library authors - maybe we still want `pd.datetime()` to be backed by our datetime extension by default, but could then use pyarrow's date implementation to back `pd.date()` by default rather than having to build that from scratch

In today's world we are either asking users to explicitly use `pa.timestamp()` to get back a true date type from `ser.dt.date`, or they use the traditional pandas extension type and get back `dtype=object` when they use the accessor

Reply:
> This PDEP suggests the 99% use case would be pd.datetime(tz=tz)

This seems reasonable, but then im not clear on why physical_type is in the base class.

> but could then use pyarrow's date implementation to back pd.date() by default

Doesn't this then give the user mixed-and-matched propagation behavior that we agreed was unwanted?

Reply:
> I think it would be easier to discuss if there were more clarity on what this base class is used for. With that very big caveat, my impression is that the relevant keyword/attribute will vary between string/decimal/datetime/whatever and this doesn't belong in the base class at all.

Are you referring to the `physical_type` keyword or something else? Every logical type instance should have a physical type underneath. So the "string" type could be physically implemented as a `pa.string()`, `pa.large_string()`, `object`, etc...The "datetime" type could be physically implemented as a `pa.timestamp` or the existing `pandas extension datetime`. Right now a decimal would only be backed by either a `pa.decimal128` or `pa.decimal256`


> If i have a non-pyarrow pd.datetime Series and i do ser.dt.date and get a pd.date Series that under the hood is date32[pyarrow] then i have mixed semantics.

In the logical type world that is totally fine. It is up to our implementation to manage the fact that a datetime can be implemented either as a `pa.timestamp` or our own datetime type, and a date can only be implemented as a `pa.date32`, but that is not a concern of the end user. I am definitely not suggesting we develop a non-pyarrow date type


> BTW im nitpicking a bunch, but im generally positive on the central idea here.

The feedback is very helpful and its good to flesh this conversation out so no worries. Much appreciated

Reply:
While I certainly understand the idea of a physical type, I am not sure such a simple attribute will fit for all cases. For example for our own masked Int64Dtype, what's the physical type? `np.int64`? (but that's only for the data array, while there is also a mask array) 
And is this physical type then always either a numpy type or pyarrow type?

Reply:
I would consider the physical type in that case to be the `pandas Int64 extension type` but maybe we should add terminology to this PDEP? 

In the scope of this document a logical type has no control over the layout of a data buffer or mask; it simply serves as a conceptual placeholder for the idea of a type. The physical type, by contrast, does control how its data buffer and mask are laid out

Reply:
>> I think it would be easier to discuss if there were more clarity on what this base class is used for. With that very big caveat, my impression is that the relevant keyword/attribute will vary between string/decimal/datetime/whatever and this doesn't belong in the base class at all.

> Are you referring to the physical_type keyword or something else? Every logical type instance should have a physical type underneath. So the "string" type could be physically implemented as a pa.string(), pa.large_string(), object, etc...The "datetime" type could be physically implemented as a pa.timestamp or the existing pandas extension datetime. Right now a decimal would only be backed by either a pa.decimal128 or pa.decimal256

The "more clarity" request was for what we do with BaseDtype.  I guess examples for both usage and construction with, say, `pd.DatetimeTZDtype(unit="ns", tz="US/Pacific")` would also be helpful.

>> If i have a non-pyarrow pd.datetime Series and i do ser.dt.date and get a pd.date Series that under the hood is date32[pyarrow] then i have mixed semantics.

> In the logical type world that is totally fine. It is up to our implementation to manage the fact that a datetime can be implemented either as a pa.timestamp or our own datetime type, and a date can only be implemented as a pa.date32, but that is not a concern of the end user. I am definitely not suggesting we develop a non-pyarrow date type

You're saying mixed semantics are fine here, but in #58315 you agreed we should avoid that.  What changed?  Having consistent/predictable semantics is definitely a concern of the end user.

Reply:
> The "more clarity" request was for what we do with BaseDtype. I guess examples for both usage and construction with, say, `pd.DatetimeTZDtype(unit="ns", tz="US/Pacific")` would also be helpful.

I think it should be as simple as:

```python
>>> ser = pd.Series(dtype=pd.datetime(unit="ns", tz="US/Pacific")
>>> ser.dtype
pd.DatetimeDtype(unit="ns", tz="US/Pacific")
```

Or

```python
>>> ser = pd.Series(dtype=pd.DatetimeDtype(unit="ns", tz="US/Pacific")
>>> ser.dtype
pd.DatetimeDtype(unit="ns", tz="US/Pacific")
```

> You're saying mixed semantics are fine here, but in #58315 you agreed we should avoid that. What changed? Having consistent/predictable semantics is definitely a concern of the end user.

I think interoperating between pyarrow / numpy / our own extension types behind the scenes is perfectly fine. It would be problematic to have those implementation details spill out into the end user space though. 

When a user says "I want a datetime type" they should just get something that works like a datetime type. When they then say `ser.dt.date()` they should get back something that works like a date. Having logical types works as a contract with the end user to promise we will deliver that. Whether we use NumPy, pyarrow, or something else underneath to do that is no concern of the end user